specs/WebResourceRequested-CustomScheme.md

```diff
@@ -0,0 +1,414 @@
+  /// Gets the `ICoreWebView2CustomSchemeRegistration` located in the
+  /// `ICoreWebView2CustomSchemeRegistrationCollection` at the given index.
+  HRESULT GetValueAtIndex([in] UINT32 index,
+                          [out, retval] ICoreWebView2CustomSchemeRegistration** registration);
+
+  /// Inserts a `ICoreWebView2CustomSchemeRegistration` into the
+  /// `ICoreWebView2CustomSchemeRegistrationCollection` at the given index.
+  HRESULT InsertValueAtIndex(
+      [in] UINT32 index,
+      [in] ICoreWebView2CustomSchemeRegistration* value);
+
+  /// The number of `ICoreWebView2CustomSchemeRegistration`s in the collection.
+  [propget] HRESULT Count([out, retval] UINT32* value);
+}
+
+// This is the ICoreWebView2EnvironmentOptions3 interface
+[uuid(ac52d13f-0d38-475a-9dca-876580d6793e), object, pointer_default(unique)]
+interface ICoreWebView2EnvironmentOptions3 : IUnknown {
+  /// Collection of custom scheme registrations
+  [propget] HRESULT CustomSchemeRegistrations(
+      [out, retval] ICoreWebView2CustomSchemeRegistrationCollection** schemeRegistrar);
+}
+```
+
+## .NET API
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    // Represents the registration of a custom scheme with the CoreWebView2Environment.
+    // This allows the WebView2 app to be able to handle
+    // WebResourceRequested event for requests with the specified scheme and
+    // be able to navigate the WebView2 to the custom scheme. Once the environment
+    // is created, the registrations are valid and immutable throughout the
+    // lifetime of the associated WebView2s' browser process and any WebView2
+    // environments sharing the browser process must be created with identical
+    // custom scheme registrations, otherwise the environment creation will fail.
+    // Any further attempts to register the same scheme will fail.
+    // The URIs of registered custom schemes will be treated similar to http URIs for their origins.
+    // They will have tuple origins for URIs with host and opaque origins for
+    // URIs without host as specified in https://html.spec.whatwg.org/multipage/origin.html
+    // Example:
+    // custom-scheme-with-host://hostname/path/to/resource has origin of custom-scheme-with-host://hostname
+    // custom-scheme-without-host:path/to/resource has origin of custom-scheme-without-host:path/to/resource
+    // For WebResourceRequested event, the cases of request URIs and filter URIs with custom schemes
+    // will be normalized according to generic URI syntax rules. Any non-ASCII characters will be preserved.
+    runtimeclass CoreWebView2CustomSchemeRegistration
+    {
+        // Constructor
+        CoreWebView2CustomSchemeRegistration(string schemeName, bool isSecure, bool isCorsEnabled, IList<String> allowedOrigins);
+
+        // The name of the custom scheme to register.
+        string SchemeName { get; set; };
+
+        // Whether the scheme will treated as a secure context like https.
```

Comment:
Can you add a link to the HTML spec or elsewhere that can define secure context?

```suggestion
        // Whether the scheme will be treated as a secure context like https. 
```